packages/next/shared/lib/match-remote-pattern.ts

```diff
@@ -0,0 +1,68 @@
+import type { RemotePattern } from './image-config'
+
+export function matchRemotePattern(pattern: RemotePattern, url: URL): boolean {
+  if (pattern.protocol !== undefined) {
+    const actualProto = url.protocol.slice(0, -1)
+    if (pattern.protocol !== actualProto) {
+      return false
+    }
+  }
+  if (pattern.port !== undefined) {
+    if (pattern.port !== url.port) {
+      return false
+    }
+  }
+  if (pattern.pathname !== undefined) {
+    const patternParts = pattern.pathname.split('/')
+    const actualParts = url.pathname.split('/')
+    const len = Math.max(patternParts.length, actualParts.length)
+    for (let i = 0; i < len; i++) {
+      if (patternParts[i] === '**' && actualParts[i] !== undefined) {
+        // Double asterisk means "match everything until the end of the path"
+        // so we can break the loop early
+        break
+      }
+      if (patternParts[i] === '*') {
+        // Single asterisk means "match this part" so we can
+        // continue to the next part of the loop
+        continue
+      }
+      if (patternParts[i] !== actualParts[i]) {
+        return false
+      }
+    }
+  }
+
+  if (pattern.hostname !== undefined) {
+    const patternParts = pattern.hostname.split('.').reverse()
+    const actualParts = url.hostname.split('.').reverse()
+    const len = Math.max(patternParts.length, actualParts.length)
+    for (let i = 0; i < len; i++) {
+      if (patternParts[i] === '**' && actualParts[i] !== undefined) {
```

Comment:
Do this mean you can't match an unspecified amount of segments and then make sure it starts with a specific value e.g. `bottom-subdomain.**.example.com`?

Would be good to validate/document this as well if so.

Reply:
Correct. Can you think of a use case where that would be necessary?

Reply:
Can't think of a specific case at the moment it'd be needed and should be an ok limitation mainly mentioning for validating the input doesn't expect this/documenting it

Reply:
I added an error in this case and documented it is not allowed 👍 

We'll see if users need something like this before we change from experimental to stable.